tools/clustering/src/clustering.py

```diff
@@ -0,0 +1,269 @@
+
+
+def estimate_dbscan_epsilon(trace, coverage):
+
+    assert 0 < coverage <= 1, "Coverage should be within the (0,1] range."
+
+    n = trace.get_invocation_count()
+
+    print("Finding eps parameter based on coverage of %f..." % coverage)
+    distance_matrix = trace.get_distance_matrix(disjoint_sets)
+    distances = []
+    for i in range(n):
+        distances.append(min([distance_matrix[i, j] for j in range(n) if j != i]))
+
+    pointLimit = math.ceil((n - 1) * coverage)
+    distances = np.sort(distances)
+    uniquedist = np.unique(distances)
+    distanceLimit = distances[pointLimit]
+    nextDist = uniquedist[1 + np.where(uniquedist == distanceLimit)[0]]
+
+    print("eps parameter based on coverage of %f set to %f" % (coverage, (distanceLimit + nextDist)/2))
+    return (distanceLimit + nextDist)/2
+
+def dbscan(trace, epsilon):
+    distance_matrix = trace.get_distance_matrix(disjoint_sets)
+
+    print("Clustering using parameters: eps = %f;" % epsilon)
+    db = DBSCAN(metric="precomputed", eps=epsilon).fit(distance_matrix)
+    core_samples_mask = np.zeros_like(db.labels_, dtype=bool)
+    core_samples_mask[db.core_sample_indices_] = True
+    labels = db.labels_
+
+    # Number of clusters in labels, ignoring noise if present.
+    n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
+    n_noise_ = list(labels).count(-1)
+
+    print('Estimated number of clusters: %d' % n_clusters_)
+    print('Estimated number of noise points: %d (%f%%)' % (n_noise_, n_noise_*100/trace.get_invocation_set_count()))
+    if n_clusters_ > 1:
+        silhouette_score = metrics.silhouette_score(distance_matrix, labels,
+                                                    metric="precomputed")
+        print("Silhouette Coefficient: %0.3f" % silhouette_score)
+
+    clusters = [[] for i in range(n_clusters_)]
+    noise_invocations = []
+    for i in range(len(labels)):
+        label = labels[i]
+
+        if label == -1:
+            noise_invocations.append(i)
+        else:
+            clusters[label].append(i)
+
+    invocation_sets = [invocation_set.invocations for invocation_set in trace.invocation_sets]
+
+    return ClusteringInformation(epsilon, invocation_sets, clusters, noise_invocations)
+
+def dbscan_ipc_instr(invocations, epsilon):
+    data = []
+    for i in range(len(invocations)):
+        invocation = invocations[i]
+        done = False
+        for dp in data:
+            if not done and abs(1 - (invocation.metrics.instructions / dp[1])) < 0.01:
+                count = len(dp[0])
+                dp[0].append(invocation)
+                dp[1] = ((dp[1] * count) + invocation.metrics.instructions) / (count + 1)
+                dp[2] = ((dp[2] * count) + invocation.metrics.ipc) / (count + 1)
+                dp[3].append(i)
+                done = True
+        if done == False:
+            data.append([[invocation], invocation.metrics.instructions, invocation.metrics.ipc, [i]])
```

Comment:
@ArnauBigas Please explain this peace of code and why is performed. If I understand correctly, seems to me that you _pre-cluster_ the invocations by instruction count?  is this done to speedup the following clustering step?

Reply:
That is correct, sklearn can't easily deal with clustering some of the millions of invocations found in some benchmarks. Also, keep in mind that some of the machines we use in our use-case are relatively weak ;)

So, yeah, these are optimizations to speed-up the clustering.

Reply:
Now, I get it. Note that the optimization _prioritizes_  the #instructions metric since it _averages_ and _joins_ invocations with similar #instructions that might have quite different IPC. This is not a problem as we have seen experimentally. In fact, I guess that that rarely happens (having the same #instructions and significantly different IPC) .  Intuition tells me that the same #instructions indicate that the same code path has been executed, and therefore we should expect the same performance.
That makes me wonder, have we ever tried to use #instructions alone as a clustering approach similar to what we did for IPC alone? If that turns out to be as good as this approach, it will be better since the #instruction metric is much less prone to measurement/experimental noise.

Reply:
Yes, you understood that correctly. The argument for prioritizing the #instructions is the one I used when I wrote that bit of code. Similar amount of instructions has a high probability of being the same execution path, although coincidentally two different code paths could have the same #instructions.

These are the clustering ideas we have used so far:
1. Clustering based on #page accesses and the address of each access.
2. Clustering based on IPC (prone to the noise you mention)
3. Clustering based on IPC & #instructions (the one we're using right now)

So, no, we haven't used instructions alone ;)